### PR TITLE
Autocomplete

### PR DESCRIPTION
**Pull request by ozerovs**

### CODE REVIEW

File: preview/containers/content/index.es [hunk 1]
```diff
@@ -56,6 +82,20 @@ class Inputs extends PureComponent {
     console.warn('Selected emoji:', list[key]);
   }
 
+  handleTagsPickerSelect(tag) {
+    const value = this.inputRef.current.getValue().concat([tag]);
```
Comment (atomiomi) [inline preview/containers/content/index.es]: why do u need this?

File: preview/containers/content/index.es [hunk 2]
```diff
@@ -56,6 +82,20 @@ class Inputs extends PureComponent {
...
+    const stateUpdate = (prevState) => ({ extraItems: prevState.extraItems.concat([tag]) });
+    const callback = () => this.setState(stateUpdate);
+
+    this.inputRef.current.setValue(value, callback);
+  }
+
+  handleAutocomplete(value) {
+    console.info('Got autocomplete input:', value);
+    const autocompleteSuggestions = (value && value.length) ? Array.from(value) : [];
```
Comment (atomiomi) [inline preview/containers/content/index.es]: value.split('')

File: src/autocomplete/index.es [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+    if (options !== prevProps.options) {
+      if (options && options.length) this.show();
+      else this.hide();
+    }
+  }
+
+  componentWillUnmount() {
+    this.hide();
+    this.isComponentMounted = false;
```
Comment (atomiomi) [inline src/autocomplete/index.es]: you can create prop on unmount only `isUnmounted = true`.

File: src/autocomplete/index.es [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+  }
+
+  setEl(name) {
+    return (el) => { this.els[name] = el; };
```
Comment (atomiomi) [inline src/autocomplete/index.es]: maybe createRef is better for this case

File: src/constants/misc.es [hunk 5]
```diff
@@ -0,0 +1 @@
+export const TAG_LIMIT = 34;
```
Comment (atomiomi) [inline src/constants/misc.es]: put it in tag cloud component

File: src/constants/times.es [hunk 6]
```diff
@@ -0,0 +1 @@
+export const SEARCH_CHANGE_RATE = 300;
```
Comment (atomiomi) [inline src/constants/times.es]: put it in tag cloud component

File: src/tag_cloud/index.es [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    });
+
+    let error;
+    if (this.isErrorActive()) error = <em className="ui-error">{this.getError()}</em>;
+
+    let labelNode;
+    if (label) labelNode = <strong className="ui-label c-tag_cloud-label">{label}</strong>;
+
+    return (
+      <article className={className} ref={this.setEl('container')}>
```
Comment (atomiomi) [inline src/tag_cloud/index.es]: input component has setEl method?

File: src/tags_picker/index.es [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+class TagsPicker extends PureComponent {
+  constructor(props) {
+    super(props);
+    bindTo(this,
+      'handleSelect',
+    );
+    this.els = {};
+  }
+
+  setEl(name) {
```
Comment (atomiomi) [inline src/tags_picker/index.es]: better to use createRef

File: preview/containers/content/index.es (hunk 1 above)
Comment (ozerovs) [reply preview/containers/content/index.es]: For a proper example of tag cloud population with new tags.

File: src/tag_cloud/index.es (hunk 7 above)
Comment (ozerovs) [reply src/tag_cloud/index.es]: Yes

File: preview/containers/content/index.es (hunk 1 above)
Comment (atomiomi) [reply preview/containers/content/index.es]: no need, for tag cloud receives only array of items.

File: package.json [hunk 9]
```diff
@@ -7,7 +7,7 @@
   "homepage": "https://nebenan.de/",
   "repository": "good-hood-gmbh/nebenan-components",
   "bugs": "https://github.com/good-hood-gmbh/nebenan-components/issues",
-  "version": "8.3.2",
+  "version": "8.3.3",
```
Comment (atomiomi) [inline package.json]: version 8.5.0

File: src/tag_cloud/index.scss [hunk 10]
```diff
@@ -0,0 +1,19 @@
+.c-tag_cloud {
+  &.is-locked .c-tag_cloud-item {
+    cursor: default;
+    opacity: .7;
```
Comment (atomiomi) [inline src/tag_cloud/index.scss]: check style linter, probably it will required 0.7 format

File: src/autocomplete/index.es [hunk 11]
```diff
@@ -155,6 +157,7 @@ Autocomplete.propTypes = {
   ]),
   getOption: PropTypes.func,
   listChildren: PropTypes.node,
+  listTopChildren: PropTypes.node,
```
Comment (suprMax) [inline src/autocomplete/index.es]: listHeader
listFooter

or pass a render prop (even better)?

File: preview/containers/content/index.es [hunk 12]
```diff
@@ -233,11 +241,18 @@ class Inputs extends PureComponent {
 
         <div className="preview-section">
           <Autocomplete
-            label="Autocomplete" name="autocomplete"
+            label="Default Autocomplete" name="autocomplete"
             placeholder="2, 140" options={this.state.suggestions}
             onInput={this.handleAutocomplete}
             error="Required between 2 and 140 chars" required
           />
+          <Autocomplete
```
Comment (atomiomi) [inline preview/containers/content/index.es]: leave only one component

File: preview/containers/content/index.es [hunk 13]
```diff
@@ -233,11 +241,18 @@ class Inputs extends PureComponent {
...
+            label="Autocomplete with non-default list" name="autocomplete"
+            placeholder="Type in me" options={this.state.suggestions2}
+            onInput={this.handleAutocomplete2}
+            error="Required between 2 and 140 chars" required
+            renderChildList={(data) => <div style={{ background: 'lightgreen' }}>{data}</div>}
```
Comment (atomiomi) [inline preview/containers/content/index.es]: renderList

Comment (atomiomi) [reply preview/containers/content/index.es]: data -> listComponent

File: src/autocomplete/index.es [hunk 14]
```diff
@@ -101,25 +101,27 @@ class Autocomplete extends PureComponent {
...
     return (
       <div className="c-autocomplete-content ui-card">
-        {listHeader}
-        <ContextList
-          ref={this.list} className="ui-options"
-          options={options} getOption={getOption}
-          onSelect={this.handleSelect}
-        />
-        {listFooter}
+        {renderChildList ? renderChildList(contextList) : contextList}
```
Comment (atomiomi) [inline src/autocomplete/index.es]: move to variable

**Pull request merged**